File: cogs/utility/autoreact_cog.py
import discord
from discord.ext import commands
from utils.rate_limiter import RateLimiter
from utils.permissions import has_permissions_to_use_external_emojis
import asyncio

class AutoReactCog(commands.Cog):
    def __init__(self, bot):
        self.bot = bot
        self.user_reactions = {}
        self.rate_limiter = RateLimiter()

    @commands.command(name='react')
    @has_permissions_to_use_external_emojis()
    async def react(self, ctx, action: str, user_ids: commands.Greedy[int] = None, *emojis):
        if action.lower() not in ['enable', 'disable', 'list']:
            await ctx.send("Invalid action. Please choose 'enable', 'disable', or 'list'.")
            return

        if action.lower() == 'list':
            await self.list_reactions(ctx)
            return

        if not user_ids:
            user_ids = [ctx.author.id]

        for user_id in user_ids:
            if action.lower() == 'enable':
                await self.enable_reactions(ctx, user_id, emojis)
            elif action.lower() == 'disable':
                await self.disable_reactions(ctx, user_id)

    async def list_reactions(self, ctx):
        if self.user_reactions:
            response = "Auto-react enabled for the following users:\n"
            for user_id, emojis in self.user_reactions.items():
                response += f"User ID: {user_id}, Emojis: {', '.join(emojis)}\n"
            await ctx.send(f"```{response}```")
        else:
            await ctx.send("No users have auto-react enabled.")

    async def enable_reactions(self, ctx, user_id, emojis):
        if user_id not in self.user_reactions:
            self.user_reactions[user_id] = set()
        self.user_reactions[user_id].update(emojis)
        await ctx.send(f"Auto-react enabled for user: {user_id} with emojis: {', '.join(emojis)}")

    async def disable_reactions(self, ctx, user_id):
        if user_id in self.user_reactions:
            del self.user_reactions[user_id]
            await ctx.send(f"Auto-react disabled for user: {user_id}")

    @commands.Cog.listener()
    async def on_message(self, message):
        if message.author.id in self.user_reactions:
            for emoji in self.user_reactions[message.author.id]:
                await self.rate_limiter.wait()
                await message.add_reaction(emoji)

def setup(bot):
    bot.add_cog(AutoReactCog(bot))

File: cogs/utility/autoreact_cog.ts
import { Client, Message, GuildMember } from 'discord.js-selfbot-v13';
import { Command } from 'discord-akairo';
import { RateLimiter } from '../../utils/rate_limiter';
import { hasPermissionsToUseExternalEmojis } from '../../utils/permissions';
import { enableReactions, disableReactions, listReactions } from '../../utils/autoreact_utils';

class AutoReactCog {
    private client: Client;
    private userReactions: { [key: string]: Set<string> };
    private rateLimiter: RateLimiter;

    constructor(client: Client) {
        this.client = client;
        this.userReactions = {};
        this.rateLimiter = new RateLimiter();
    }

    @Command({
        aliases: ['react'],
        description: 'Enable, disable, or list auto-react for specified users with given emojis.',
        args: [
            { id: 'action', type: 'string' },
            { id: 'userIds', type: 'string', match: 'rest', default: null },
            { id: 'emojis', type: 'string', match: 'rest', default: null }
        ]
    })
    @hasPermissionsToUseExternalEmojis()
    async react(message: Message, { action, userIds, emojis }: { action: string, userIds: string | null, emojis: string | null }) {
        if (!['enable', 'disable', 'list'].includes(action.toLowerCase())) {
            await message.channel.send("Invalid action. Please choose 'enable', 'disable', or 'list'.");
            return;
        }

        if (action.toLowerCase() === 'list') {
            await listReactions(message, this.userReactions);
            return;
        }

        const userIdArray = userIds ? userIds.split(' ').map(id => id.trim()) : [message.author.id];
        const emojiArray = emojis ? emojis.split(' ').map(emoji => emoji.trim()) : [];

        for (const userId of userIdArray) {
            if (action.toLowerCase() === 'enable') {
                await enableReactions(message, userId, emojiArray, this.userReactions);
            } else if (action.toLowerCase() === 'disable') {
                await disableReactions(message, userId, this.userReactions);
            }
        }
    }

    @Client.on('message')
    async onMessage(message: Message) {
        try {
            if (this.userReactions[message.author.id]) {
                for (const emoji of this.userReactions[message.author.id]) {
                    await this.rateLimiter.wait();
                    await message.react(emoji);
                }
            }
        } catch (error) {
            console.error(`Failed to react to message. Error: ${error.message}`);
        }
    }
}

export function setup(client: Client) {
    return new AutoReactCog(client);
}
